src/sofa/pbrpc/web_service.cc

```diff
@@ -0,0 +1,539 @@
+    RegisterServlet("/services", _default_services);
+
+    _default_service = 
+        sofa::pbrpc::NewPermanentExtClosure(this, &WebService::DefaultService);
+    RegisterServlet("/service", _default_service);
+}
+
+bool WebService::RegisterServlet(const std::string& path, Servlet servlet)
+{
+    std::string real_path = FormatPath(path);
+    ScopedLocker<FastLock> _(_servlet_map_lock);
+    if (!_servlet_map.unique())
+    {
+        ServletMapPtr servlet_map(new ServletMap(*_servlet_map));
+        _servlet_map.swap(servlet_map);
+    }
+    std::pair<ServletMap::iterator, bool> ret = 
+        _servlet_map->insert(std::make_pair(real_path, servlet));
+    if (ret.second == true)
+    {
+#if defined( LOG )
+        LOG(INFO) << "RegisterServlet(): register webserver path {" 
+            << path << "} format to {" << real_path << "} success";
+#else
+        SLOG(INFO, "RegisterServlet(): register webserver path {%s} format to {%s} success", 
+             path.c_str(), real_path.c_str());
+#endif
+    }
+    else
+    {
+#if defined( LOG )
+        LOG(ERROR) << "RegisterServlet(): register webserver path {" 
+            << path << "} format to {" << real_path << "} alreay exist";
+#else
+        SLOG(ERROR, "RegisterServlet(): register webserver path {%s} format to {%s} already exist", 
+             path.c_str(), real_path.c_str());
+#endif
+    }
+    return ret.second;
+}
+
+bool WebService::UnregisterServlet(const std::string& path)
+{
+    std::string real_path = FormatPath(path);
+    ScopedLocker<FastLock> _(_servlet_map_lock);
+    if (!_servlet_map.unique())
+    {
+        ServletMapPtr servlet_map(new ServletMap(*_servlet_map));
+        _servlet_map.swap(servlet_map);
+    }
+    ServletMap::iterator it = _servlet_map->find(real_path);
+    if (it == _servlet_map->end())
+    {
+#if defined( LOG )
+        LOG(ERROR) << "UnregisterServlet(): unregister webserver path {" 
+            << path << "} format to {" << real_path << "} not exist";
+#else
+        SLOG(ERROR, "UnregisterServlet(): unregister webserver path {%s} format to {%s} not exist", 
+             path.c_str(), real_path.c_str());
+#endif
+        return false;
+    }
+    _servlet_map->erase(it);
+#if defined( LOG )
+    LOG(INFO) << "UnregisterServlet(): unregister webserver path {" 
+        << path << "} format to {" << real_path << "} success";
+#else
+    SLOG(INFO, "UnregisterServlet(): unregister webserver path {%s} format to {%s} success", 
+         path.c_str(), real_path.c_str());
+#endif
+    return true;
+}
+
+bool WebService::RoutePage(
+    const RpcRequestPtr& rpc_request,
+    const RpcServerStreamWPtr& server_stream)
+{
+    const HTTPRpcRequest* http_rpc_request = 
+        static_cast<const HTTPRpcRequest*>(rpc_request.get());
+
+    HTTPRequest request;
+    request.path = http_rpc_request->_path;
+    request.body = http_rpc_request->_req_body->ToString();
+    request.headers = http_rpc_request->_headers;
```

Comment:
consider use "const std::map*" to pass headers and query_params? this would avoid unnecessary copy.